snow/engine/snowman/transitive_test.go

```diff
@@ -2639,6 +2670,276 @@ func TestEngineRepollsMisconfiguredSubnet(t *testing.T) {
 	require.Equal(choices.Accepted, blk.Status())
 }
 
+// Full blockchain structure:
+//
+//	  G
+//	 / \
+//	0   3
+//	|   |
+//	1   4
+//	|
+//	2
+//
+// K = 3, Alpha = 2, Beta = 1, ConcurrentRepolls = 1
+//
+// Initial configuration:
+//
+//	G
+//	|
+//	0
+//	|
+//	1
+//	|
+//	2
+//
+// The following is a regression test for a bug where the engine would stall.
+//
+//  1. Poll = 0: Handle a chit for block 1.
+//  2. Poll = 0: Handle a chit for block 2.
+//  3. Poll = 0: Handle a chit for block 3. This will issue a Get request for block 3. This will block on the issuance of block 3.
+//  4. Attempt to issue block 4. This will block on the issuance of block 3.
+//  5. Poll = 1: Handle a chit for block 1.
+//  6. Poll = 1: Handle a chit for block 2.
+//  7. Poll = 1: Handle a chit for block 4. This will block on the issuance of block 4.
+//  8. Issue block 3.
+//     Poll = 0 terminates. This will accept blocks 0 and 1. This will also reject block 3.
+//     Block = 4 will attempt to be delivered, but because it is effectively rejected due to the acceptance of block 1, it will be dropped.
+//     Poll = 1 should terminate and block 2 should be repolled.
+func TestEngineVoteStallRegression(t *testing.T) {
```

Comment:
This is basically a "black box" test of the engine... I wanted to avoid touching the internals (to avoid future maintenance burden).

We could additionally add a "clear box" test of `deliver` as a very targeted test.

Reply:
Yeah... The engine code is probably always going to have this job queue logic (So there will be some state between calls)... But I do hope that we will be able to significantly improve the testability + readability of the engine in some future PRs.

What I really want this to look like at the end of the day is specifying the "initial" state of the engine (with pending jobs and pending polls) and then verify the "output" state of the engine after some even happens... But that's really a large refactor (which is ongoing, but not for this PR)